src/lib/hpke/DY.Lib.HPKE.Spec.fst

```diff
@@ -0,0 +1,67 @@
+module DY.Lib.HPKE.Spec
+
+open Comparse
+open DY.Core
+open DY.Lib.Comparse.Glue
+open DY.Lib.Comparse.Parsers
+
+#set-options "--fuel 0 --ifuel 1"
+
+/// This module defines the specification of HPKE.
+/// More precisely, this specification corresponds to the single-shot API in base mode
+/// (c.f. section 6.1 of RFC 9180 [1]).
+///
+/// This specification has the following limitations:
+/// - the message formatting is not compliant with RFC 9180 (e.g. the one in LabeledExpand)
+/// - some KDF.Extract were omitted, for example:
+///   - `info` is extracted before going to LabeledExpand's info
+///   - `shared_secret` is extracted with the empty PSK
+///      before going to LabeledExpand's prk
+///   (c.f. section 5.1 of RFC 9180 [2])
+/// Having a more compliant specification is left as future work.
+///
+/// [1] https://www.rfc-editor.org/rfc/rfc9180.html#name-encryption-and-decryption-2
+/// [2] https://www.rfc-editor.org/rfc/rfc9180.html#name-creating-the-encryption-con
+
+/// Definition of the internal function LabeledExpand
+
+[@@ with_bytes bytes]
+type hpke_labeled_expand_info = {
+  len: nat_lbytes 2;
+  [@@@ with_parser #bytes ps_string]
+  label: string;
+  info: bytes;
+}
+
+%splice [ps_hpke_labeled_expand_info] (gen_parser (`hpke_labeled_expand_info))
+%splice [ps_hpke_labeled_expand_info_is_well_formed] (gen_is_well_formed_lemma (`hpke_labeled_expand_info))
+
+instance parseable_serializeable_bytes_hpke_labeled_expand_info: parseable_serializeable bytes hpke_labeled_expand_info =
+  mk_parseable_serializeable ps_hpke_labeled_expand_info
+
+val hpke_labeled_expand: bytes -> string -> bytes -> len:nat_lbytes 2{len <> 0} -> bytes
+let hpke_labeled_expand prk label info len =
+  kdf_expand prk (serialize _ { len; label; info }) len
+
+/// The HPKE functions
+
+val hpke_pk: bytes -> bytes
+let hpke_pk sk =
+  kem_pk sk
+
+val hpke_enc: bytes -> bytes -> bytes -> bytes -> bytes -> (bytes & bytes)
+let hpke_enc pkR entropy plaintext info ad =
+  let (enc, shared_secret) = kem_encap pkR entropy in
+  let aead_key = hpke_labeled_expand shared_secret "key" info 32 in
+  let aead_nonce = hpke_labeled_expand shared_secret "nonce" info 32 in
```

Comment:
These strings are the one used in the HPKE RFC (9180), I think they shouldn't be changed. Hopefully this doesn't bring conflicts with other protocols, because `shared_secret` has a distinct label (after implementing your previous remark, "DY.Lib.HPKE")

Reply:
Ah, actually [HPKE is using `base_nonce`](https://www.rfc-editor.org/rfc/rfc9180.html#section-5.1-10). I will change that.

Reply:
Changed to `"base_nonce"`

Reply:
Do I understand correctly that these labels are never used to split crypto predicates? In this case we can leave them as they are.

Reply:
While I think it makes sense to use the strings from the RFC, it might make sense to pull them out into some global constant, just so that if updates are needed (like the recent change from `"nonce"` to `"base_nonce"`), it's easy to see that it changed everywhere correctly.

I guess a variable may run into issues when trying to match with this constant string, but maybe there's some way around that.

Reply:
@fabian-hk
It is not currently not used to split crypto predicates, the splitting only happens on the `KdfExpandKey` usage tag, which is `"DY.Lib.HPKE"`, meaning that we define local `kdf_expand` invariants for HPKE, in `hpke_kdf_expand_usage`. This local invariant is currently defined monolithically, but we could also do predicate splitting on it, by splitting on the label (either `"key"` or `"base_nonce"`), and that would not interfer with other protocols as it would only be relevant for keys with usage tag `KdfExpandKey "DY.Lib.HPKE" ...`.

I don't think it's worth doing this additional level of splitting though, it wouldn't improve proofs dramatically but it would add a few hundred lines of code (to instantiate the split function mechanism on it).

@qaphla It seems like better coding practice for sure, but I'm not sure it would be worth here, as the strings are not meant to change (now that they are correct with respect to the RFC), they are short (hence the new variable names would be longer), and we can nicely pattern-match on them (otherwise in the predicates we would need an additional sequence of `if`s). For these reasons I would prefer keeping it as-is, but if you still think it's better to introduce global constants for them I can do that.

Reply:
I think I agree for now, but I might see if the F* devs have any plans to have a feature like C's `#define`, allowing for something that better captures the idea of a global constant, as compared to a static global variable. If such a feature existed in F*, I think it would still be worth using even with these short names, because it would solve the problem of pattern matching. Indeed, that would be useful for most cases where we're working with usage strings or other similar string tags.

Reply:
That would be a good usecase for a `#define`-like construct, but it doesn't seem that popular in languages modern than C!